compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -364,14 +365,46 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
           }
         }
 
-        val clDefs = generatedClasses.toList
+        val clDefs = if (generatedStaticForwarderClasses.isEmpty) {
+          /* Fast path, applicable under -Xno-forwarders, as well as when all
+           * the `object`s of a compilation unit have a companion class.
+           */
+          generatedClasses.toList
+        } else {
+          val regularClasses = generatedClasses.toList
+
+          /* #4148 Add generated static forwarder classes, except those that
+           * would collide with regular classes on case insensitive file
+           * systems.
+           */
+
+          /* I could not find any reference anywhere about what locale is used
+           * by case insensitive file systems to compare case-insensitively.
+           * In doubt, force the English locale, which is probably going to do
+           * the right thing in virtually all cases (especially if users stick
+           * to ASCII class names), and it has the merit of being deterministic,
+           * as opposed to using the OS' default locale.
+           */
+          def caseInsensitiveNameOf(classDef: js.ClassDef): String =
+            classDef.name.name.nameString.toLowerCase(java.util.Locale.ENGLISH)
+
+          val generatedCaseInsensitiveNames =
+            regularClasses.map(caseInsensitiveNameOf).toSet
+          val staticForwarderClasses = generatedStaticForwarderClasses.toList.filter { classDef =>
+            !generatedCaseInsensitiveNames.contains(caseInsensitiveNameOf(classDef))
+          }
```

Comment:
Can we at least warn (just like scalac) when we are dropping static forwarders?

Reply:
We can, but then we'd have to propose a flag to disable that warning, or we're going to make some library maintainers angry at us for breaking -Xfatal-warnings in a way that is not recoverable without breaking their backward binary compat.

Note that scalac never drops static forwarders for the reason that we do here. Instead if *will* erase things, and will emit a warning that the result will be broken on case insensitive file systems. However it only emits static forwarders for top-level objects, not for static objects inside objects. Here, we are instead fixing the potential clash. The result is more correct than what scalac lets through for top-level objects, so not necessarily as deserving of a warning as the case (mis-)handled by scalac.

Reply:
My main concern is that reasoning about the existence of static forwarders (from a users POV) becomes non-local and difficult.

In general, I find non-opt-in do-it-if-you-can behavior has a high potential to cause confusing behavior. I am not set on a warning. Maybe it is better if we protect exporting nested static forwarders with a flag (for compilation of the JDK) but hard fail if there is a conflict. IIUC, this would address all use cases we have and avoid potential confusing behavior. (IIUC, nobody needs the avoidance, it's just that static forwarders are always emitted).

Reply:
I can prepare an alternative PR with an option to opt-in for nested static forwarders. However I don't think we should hard-fail in that case. We can't hard-fail for top-level static forwarders, because scalac does not hard-fail for them, and we could break actual things that work today. But then if don't hard-fail for top-level ones, but we do hard-fail for nested ones, that's going to become very tricky to handle in the code.

Reply:
Hmmm... But is the concern about not hard-failing mostly driven by how hard it is to implement or how hard it will be to use? It seems that in any case we want our behavior for nested classes to be different from our behavior for top-level classes (because on the top-level scalac precedent is more important). But at that point, shouldn't we pick what is easiest to use and not what is easiest to build?

Reply:
Ohh... Or are we aligning our behavior for *both* cases in a way that improves on scalac's behavior? In that case, that would indeed be better. But then we're back to the warning, where I'm not sure *not* having a warning in this case is better. Maybe we should just add an annotation to silence the warning (just like `uncheckedVariance` and the like).

Reply:
IMO the ideal would be to use the code of this PR, but emit a warning every time we discard a static forwarder class, with the ability to silence that warning. Since Scala 2.13.2 ([release notes](https://github.com/scala/scala/releases/tag/v2.13.2)), there is a mechanism for configurable warnings, that we could/should integrate with so that users can silence those warnings if they want. But that leaves a problem with older versions of Scala, where that's not possible. In the past we had "transitioning" compiler plugin options to selectively silence some kinds of warnings (notably related to all the Export-related features we deprecated in 0.6.17 or so). Perhaps we could introduce another one like that.

Reply:
I've added a commit with a warning that is configurable with the Scala 2.13.2+ mechanism. On previous versions of Scala, it is not configurable, however.

Reply:
I think this is good. That will allow folks to use `-Xfatal-warnings` at least on 2.13. If somebody really wants to disable it, we can still add a compiler flag for other versions.

Can we have a compiler test for it though?

Reply:
It turns out that this break 3 partest tests. I'm starting to think that this thing is much more widespread than I thought. So in addition of the warning (for which I'll add a test), I'll add a Scala.js option to opt-in to generating the static forwarders of non-top-level objects, as you suggested earlier.